src/parquet/file/metadata-internal.cc

```diff
@@ -0,0 +1,525 @@
+  return impl_->num_columns();
+}
+
+int64_t RowGroupMetaData::num_rows() const {
+  return impl_->num_rows();
+}
+
+int64_t RowGroupMetaData::total_byte_size() const {
+  return impl_->total_byte_size();
+}
+
+std::unique_ptr<ColumnMetaData> RowGroupMetaData::GetColumnMetaData(int i) const {
+  return impl_->GetColumnMetaData(i);
+}
+
+// file metadata
+class FileMetaData::FileMetaDataImpl {
+ public:
+  explicit FileMetaDataImpl(const uint8_t* metadata)
+      : metadata_{std::unique_ptr<const format::FileMetaData>(
+            reinterpret_cast<const format::FileMetaData*>(metadata))} {
+    schema::FlatSchemaConverter converter(
+        &metadata_->schema[0], metadata_->schema.size());
+    schema_.Init(converter.Convert());
+  }
+  ~FileMetaDataImpl() {}
+  int64_t num_rows() const { return metadata_->num_rows; }
+  int num_row_groups() const { return metadata_->row_groups.size(); }
+  int32_t version() const { return metadata_->version; }
+  const std::string& created_by() const { return metadata_->created_by; }
+  int num_schema_elements() const { return metadata_->schema.size(); }
+
+  void WriteTo(OutputStream* dst) {
+    SerializeThriftMsg(metadata_.get(), sizeof(format::FileMetaData), dst);
+  }
+
+  std::unique_ptr<RowGroupMetaData> GetRowGroupMetaData(int i) {
+    DCHECK(i < num_row_groups())
+        << "The file only has " << num_row_groups()
+        << " row groups, requested metadata for row group: " << i;
+    return RowGroupMetaData::GetRowGroupMetaData(
+        reinterpret_cast<const uint8_t*>(&metadata_->row_groups[i]));
+  }
+
+  const SchemaDescriptor* schema() const { return &schema_; }
+
+ private:
+  std::unique_ptr<const format::FileMetaData> metadata_;
+  SchemaDescriptor schema_;
+};
+
+std::unique_ptr<FileMetaData> FileMetaData::GetFileMetaData(const uint8_t* metadata) {
+  return std::unique_ptr<FileMetaData>(new FileMetaData(metadata));
+}
+
+FileMetaData::FileMetaData(const uint8_t* metadata)
+    : impl_{std::unique_ptr<FileMetaDataImpl>(new FileMetaDataImpl(metadata))} {}
+FileMetaData::~FileMetaData() {}
+
+std::unique_ptr<RowGroupMetaData> FileMetaData::GetRowGroupMetaData(int i) const {
+  return impl_->GetRowGroupMetaData(i);
+}
+
+int64_t FileMetaData::num_rows() const {
+  return impl_->num_rows();
+}
+
+int FileMetaData::num_row_groups() const {
+  return impl_->num_row_groups();
+}
+
+int32_t FileMetaData::version() const {
+  return impl_->version();
+}
+
+const std::string& FileMetaData::created_by() const {
+  return impl_->created_by();
+}
+
+int FileMetaData::num_schema_elements() const {
+  return impl_->num_schema_elements();
+}
+
+const SchemaDescriptor* FileMetaData::schema() const {
+  return impl_->schema();
+}
+
+void FileMetaData::WriteTo(OutputStream* dst) {
+  return impl_->WriteTo(dst);
+}
+
+// MetaData Builders
+// row-group metadata
+class ColumnMetaDataBuilder::ColumnMetaDataBuilderImpl {
+ public:
+  explicit ColumnMetaDataBuilderImpl(std::shared_ptr<WriterProperties> props,
```

Comment:
can you review the `std::shared_ptr` parameters and make them all consistently pass by const reference?